emulator.c

```diff
@@ -5,10 +5,279 @@
+  update_carry_flag(cpu, false);
+  return 4; // NOLINT
+}
+
+// Call Address
+int
+CALL(i8080 *cpu, uint16_t address)
+{
+  uint16_t return_address = readRegisterPair(cpu, PC) + 3; // NOLINT
+  cpu_write_mem(cpu, cpu->sp - 1,
+                (uint8_t)((return_address & UPPER_8_BIT_MASK) >> BYTE));
+  cpu_write_mem(cpu, cpu->sp - 2,
+                (uint8_t)(return_address & LOWER_8_BIT_MASK));
+  cpu->sp -= 2;
+  writeRegisterPair(cpu, PC, address);
+  return 17; // NOLINT
+}
+
+// Double Add
+int
+DAD(i8080 *cpu, int pair)
+{
+  uint32_t result
+      = (uint32_t)(readRegisterPair(cpu, pair) + readRegisterPair(cpu, HL));
+  update_carry_flag(cpu, result > MAX_16_BIT_VALUE);
+  writeRegisterPair(cpu, HL, (uint16_t)result);
+  return 10; // NOLINT
+}
+
+// Decrement Register
+int
+DCR(i8080 *cpu, uint8_t *reg)
+{
+  update_aux_carry_flag(cpu, *reg, MAX_8_BIT_VALUE);
+  *reg -= 1;
+  update_zero_flag(cpu, *reg);
+  update_sign_flag(cpu, *reg);
+  update_parity_flag(cpu, *reg);
+  return 5; // NOLINT
+}
+
+// Increment Register Pair
+int
+INX(i8080 *cpu, int pair)
+{
+  uint16_t value = readRegisterPair(cpu, pair) + 1;
+  writeRegisterPair(cpu, pair, value);
+  return 5; // NOLINT
+}
+
+// Jump to Address
+int
+JMP(i8080 *cpu)
+{
+  writeRegisterPair(cpu, PC, getImmediate16BitValue(cpu));
+  return 10; // NOLINT
+}
+
+// Load Accumulator
+int
+LDAX(i8080 *cpu, int pair)
+{
+  uint16_t address = readRegisterPair(cpu, pair);
+  cpu->a = cpu_read_mem(cpu, address);
+  return 7; // NOLINT
+}
+
+// Load 16-bit Data to Register Pair
+int
+LXI(i8080 *cpu, int pair, uint16_t value)
+{
+  writeRegisterPair(cpu, pair, value);
+  return 10; // NOLINT
+}
+
+// Move register value
+int
+MOV(uint8_t *dest, const uint8_t *src)
+{
+  *dest = *src;
+  return 5; // NOLINT
+}
+
+// Move register value to memory
+int
+MOV_TO_MEM(i8080 *cpu, const uint8_t *reg)
+{
+  cpu_write_mem(cpu, readRegisterPair(cpu, HL), *reg);
+  return 7; // NOLINT
+}
+
+// Move memory to register
+int
+MOV_FROM_MEM(i8080 *cpu, uint8_t *reg)
+{
+  *reg = cpu_read_mem(cpu, readRegisterPair(cpu, HL));
+  return 7; // NOLINT
+}
+
+// Move 8-bit Data to Register
+int
+MVI(uint8_t *reg, uint8_t value)
+{
+  *reg = value;
+  return 7; // NOLINT
+}
+
+// No Operation
+int
+NOP()
+{
+  return 4; // NOLINT
+}
+
+// Pop from Stack
+int
+POP(i8080 *cpu, int pair)
+{
+  uint16_t value = ((cpu_read_mem(cpu, cpu->sp + 1) << BYTE)
+                    | cpu_read_mem(cpu, cpu->sp));
+  writeRegisterPair(cpu, pair, value);
+  cpu->sp += 2;
+  return 10; // NOLINT
+}
+
+// Push to Stack
+int
+PUSH(i8080 *cpu, int pair)
+{
+  uint16_t value = readRegisterPair(cpu, pair);
+  cpu_write_mem(cpu, cpu->sp - 1,
+                (uint8_t)((value & UPPER_8_BIT_MASK) >> BYTE));
+  cpu_write_mem(cpu, cpu->sp - 2, (uint8_t)(value & LOWER_8_BIT_MASK));
+  cpu->sp -= 2;
+  return 11; // NOLINT
+}
+
+// Return to Address
+int
+RET(i8080 *cpu)
+{
+  uint16_t address = ((cpu_read_mem(cpu, cpu->sp + 1) << BYTE)
```

Comment:
Should this be sp +1 and sp + 2?

Reply:
I believe it's SP and SP+1 since it's a POP function. Takes the top two bytes off the stack and sets PC to that address.